### PR TITLE
Support aggregation in validator

### PR DESCRIPTION
**Pull request by Riateche**
- Change dependency versions to make the crate compatible with pythnet validator
- Add flags to price account to track transition from v1 to v2
- Implement transition between v1 and v2 by calling add_publisher with a special pubkey
- Skip aggregation and message buffer update in the oracle program if v2 aggregation is enabled
- Clear message buffers after transition to v2
- Expose aggregation as a library function to be used in the validator

### CODE REVIEW

File: program/rust/src/validator.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    AlreadyAggregated,
+}
+
+#[derive(Debug)]
+pub struct AggregationOutcome {
+    pub messages: [Vec<u8>; 2],
+    pub commit:   bool,
+}
+
+pub fn aggregate_price(
```
Comment (jayantk) [inline program/rust/src/validator.rs:111]: Please document this method. Few questions that the reader will have are:
1) under what conditions is `price_account_data` updated?
2) what does the `commit` flag on the return type mean?

File: program/rust/src/validator.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    unsafe {
+        c_upd_twap(price_account as *mut PriceAccount as *mut u8, agg_diff);
+    }
+
+    // We want to send a message every time the aggregate price updates. However, during the migration,
+    // not every publisher will necessarily provide the accumulator accounts. The message_sent_ flag
+    // ensures that after every aggregate update, the next publisher who provides the accumulator accounts
+    // will send the message.
+    price_account.message_sent_ = 0;
+    if price_account.update_price_cumulative().is_err() {
```
Comment (jayantk) [inline program/rust/src/validator.rs:87]: This `if` statement seems like it's a change in the logic relative to the prior implementation, and I'm not sure this is right. There are cases where aggregation can be successful and the status is unknown (e.g., not enough publisher prices are < 25 slots old).

I'm not sure what it means to "commit" here, but this seems suspicious to me.

Comment (Riateche) [reply program/rust/src/validator.rs:87]: I believe this code reproduces the original behavior. As seen [here](https://github.com/pyth-network/pyth-client/blob/main/program/rust/src/processor/upd_price.rs#L194), `upd_price` instruction handler calls `update_price_cumulative()` and passes through the error, so the transaction will revert in case `update_price_cumulative()` fails.

File: program/rust/src/validator.rs (hunk 1 above)
Comment (Riateche) [reply program/rust/src/validator.rs:122]: Added some docs.

File: program/rust/src/processor/add_publisher.rs [hunk 3]
```diff
@@ -71,6 +78,14 @@ pub fn add_publisher(
         let num_comps = try_convert::<u32, usize>(price_data.num_)?;
         sort_price_comps(&mut price_data.comp_, num_comps)?;
         return Ok(());
+    } else if cmd_args.publisher == Pubkey::from(ENABLE_ACCUMULATOR_V2) {
```
[the remark this answers is not in the document]
Comment (Riateche) [reply program/rust/src/processor/add_publisher.rs:81]: Done.

File: program/rust/src/validator.rs (hunk 2 above)
Comment (jayantk) [reply program/rust/src/validator.rs:87]: ok I missed the `?` in the `upd_price` implementation above, so you're right about this.

However, I think the if statement on line 71 above is wrong then. `c_upd_aggregate` mutates the price account content regardless of whether it returns `true` or `false`. (If it returns `false`, it sets the status flag in the account to `UNKNOWN`. If `true`, the status is `TRADING`. Either way, that change is persisted.) This update cumulative call in the original implementation is gated by the `updated` flag specifically because that guarantees that the status is `TRADING` (and thus the call won't return an error)

Comment (jayantk) [reply program/rust/src/validator.rs:87]: (as a meta-point, I believe this implementation of the function guarantees that the status flag in the price account can never be set to UNKNOWN. That's a behavior change relative to the current aggregation, so something is wrong somewhere.)

Comment (jayantk) [reply program/rust/src/validator.rs:87]: another thought here is maybe could use the existing tests of aggregation to validate that the new logic is the same as the old logic? A janky approach would be to copy/paste `test_upd_price` and then in each case there where it loads the price account to assert its contents, call this function on the PriceAccount before running the assertions.

An approach like that would give you lots of confidence in the implementation.

**Review by ali-bahjati — COMMENTED** [review]
I think as Jayant said, would be nice to test the validator update code path too.

File: program/rust/src/validator.rs (hunk 2 above)
Comment (Riateche) [reply program/rust/src/validator.rs]: > However, I think the if statement on line 71 above is wrong then. c_upd_aggregate mutates the price account content regardless of whether it returns true or false. (If it returns false, it sets the status flag in the account to UNKNOWN. If true, the status is TRADING. Either way, that change is persisted.)

Yeah, that's a good catch. I've fixed the code to not revert when `c_upd_aggregate` returns false. We'll work on adding more tests as you suggested.

> I think we should fix the behaviour on the current aggregation too.

Do you mean we should be ignoring the error returned by `update_price_cumulative()` in the contract as well? Or maybe we should change `update_price_cumulative()` to never return an error instead?

File: program/rust/src/validator.rs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    }
+    let commit = update_aggregate(slot, timestamp, price_account);
+    let messages = [
+        price_account
+            .as_price_feed_message(price_account_pubkey)
+            .to_bytes(),
+        price_account
+            .as_twap_message(price_account_pubkey)
+            .to_bytes(),
+    ];
```
Comment (ali-bahjati) [inline program/rust/src/validator.rs:139]: I think we need to add a check to not add it if the message buffer is not cleared. It can result it not sending the same account state temporarily until a new tx to update price comes but i think it's better than having duplicate messages.

File: program/rust/src/validator.rs (hunk 2 above)
Comment (ali-bahjati) [reply program/rust/src/validator.rs]: I don't have any strong opinion. I think either is fine. I rather not returning an error because it's kind of expected behaviour and cascading it can go unnoticed (like it did now)

Comment (Riateche) [reply program/rust/src/validator.rs]: Changed `update_price_cumulative()` to not return an error.

File: program/rust/src/validator.rs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+#[derive(Debug)]
+pub struct AggregationOutcome {
+    /// Messages that should be included in the merkle tree
+    /// (a price feed message and a TWAP message).
+    pub messages: [Vec<u8>; 2],
+    /// A flag indicating that the changes to `price_account_data`
+    /// should be saved to the price account.
+    pub commit:   bool,
+}
```
Comment (ali-bahjati) [inline program/rust/src/validator.rs:112]: We don't need it anymore now.

File: program/rust/src/accounts/price.rs [hunk 6]
```diff
@@ -111,17 +122,15 @@ mod price_pythnet {
             }
         }
         /// This function gets triggered when there's a succesful aggregation and updates the cumulative sums
-        pub fn update_price_cumulative(&mut self) -> Result<(), OracleError> {
+        pub fn update_price_cumulative(&mut self) {
+            debug_assert!(self.agg_.status_ == PC_STATUS_TRADING);
```
Comment (ali-bahjati) [inline program/rust/src/accounts/price.rs:126]: hmmm why would we need the debug assert here :?

Comment (Riateche) [reply program/rust/src/accounts/price.rs:126]: Our assumption is that this check should always succeed when we call `update_price_cumulative()`, no?

Comment (Riateche) [reply program/rust/src/accounts/price.rs]: Removed debug assert.

File: program/rust/src/processor/add_publisher.rs [hunk 7]
```diff
@@ -62,14 +70,19 @@ pub fn add_publisher(
...
 
-
     let mut price_data = load_checked::<PriceAccount>(price_account, cmd_args.header.version)?;
 
-    // Use the call with the default pubkey (000..) as a trigger to sort the publishers as a
-    // migration step from unsorted list to sorted list.
-    if cmd_args.publisher == Pubkey::default() {
-        let num_comps = try_convert::<u32, usize>(price_data.num_)?;
-        sort_price_comps(&mut price_data.comp_, num_comps)?;
+    if cmd_args.publisher == Pubkey::from(ENABLE_ACCUMULATOR_V2) {
```
Comment (guibescos) [inline program/rust/src/processor/add_publisher.rs:75]: Quite frankly I like a new instruction better

File: program/rust/src/accounts/price.rs [hunk 8]
```diff
@@ -111,17 +122,14 @@ mod price_pythnet {
             }
         }
         /// This function gets triggered when there's a succesful aggregation and updates the cumulative sums
-        pub fn update_price_cumulative(&mut self) -> Result<(), OracleError> {
+        pub fn update_price_cumulative(&mut self) {
```
Comment (guibescos) [inline program/rust/src/accounts/price.rs:125]: Why this change?

File: program/rust/src/validator.rs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        return Err(AggregationError::V1AggregationMode);
+    }
+
+    Ok(data)
+}
+
+fn update_aggregate(slot: u64, timestamp: i64, price_account: &mut PriceAccount) {
+    // NOTE: c_upd_aggregate must use a raw pointer to price data. We already
+    // have the exclusive mut reference so we can simply cast before calling
+    // the function.
```
Comment (guibescos) [inline program/rust/src/validator.rs:67]: imo you need `if clock.slot > latest_aggregate_price.pub_slot_` here to avoid double aggregation on the slot were the flag gets turned on

**Comment by guibescos** [comment]
I think there might be an edgecase where the price stops aggregating (both v1 and v2).
In a given slot:
- `upd_price` : aggregation is unsuccessful therefore `message_sent_ == 1`
- flags gets flipped
- aggregate v2 fails with V1AggregationMode since `PriceAccountFlags::MESSAGE_BUFFER_CLEARED` is false

Next slot:
- `upd_price` no longer triggers an aggregation because `PriceAccountFlags::ACCUMULATOR_V2` is true
- aggregate v2 also fails since `PriceAccountFlags::MESSAGE_BUFFER_CLEARED` is false

And so on...

**Comment by Riateche** [comment]
> I think there might be an edgecase where the price stops aggregating (both v1 and v2).

Good catch! I've changed the code so that the `MESSAGE_BUFFER_CLEARED` is set regardless of the `message_sent_` flag. I think this should fix the issue.

File: program/rust/src/tests/test_upd_price_with_validator.rs [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+        assert_eq!(price_data.agg_.status_, PC_STATUS_UNKNOWN);
+    }
+
+    // We aggregate the price at the end of each slot now.
+    validator::aggregate_price(1, 101, price_account.key, *price_account.data.borrow_mut())
+        .unwrap();
+    update_clock_slot(&mut clock_account, 2);
+
+    validator::aggregate_price(2, 102, price_account.key, *price_account.data.borrow_mut())
+        .unwrap();
```
Comment (ali-bahjati) [inline program/rust/src/tests/test_upd_price_with_validator.rs:128]: nit: can we once verify the output of this as well?

**Review by ali-bahjati — APPROVED** [review]
LGTM!

File: program/rust/src/processor/upd_price.rs [hunk 11]
```diff
@@ -210,76 +210,74 @@ pub fn upd_price(
...
-                        .flags
-                        .insert(PriceAccountFlags::MESSAGE_BUFFER_CLEARED);
-                }
+                        .as_price_feed_message(price_account.key)
+                        .to_bytes(),
+                    price_data.as_twap_message(price_account.key).to_bytes(),
+                ]
+            };
+
+            // Append a TWAP message if available
```
Comment (jayantk) [inline program/rust/src/processor/upd_price.rs:259]: this comment seems like it doesn't apply to anything

**Review by jayantk — APPROVED** [review]

File: program/rust/src/tests/test_upd_price_with_validator.rs (hunk 10 above)
Comment (Riateche) [reply program/rust/src/tests/test_upd_price_with_validator.rs]: Added some output checks.

File: program/rust/src/processor/upd_price.rs (hunk 11 above)
Comment (Riateche) [reply program/rust/src/processor/upd_price.rs]: Looks like it's no longer relevant. Removed.

**Review by ali-bahjati — DISMISSED** [review]

**Review by ali-bahjati — APPROVED** [review]
